okhttp-doh/src/main/java/okhttp3/doh/DnsRecordCodec.java

```diff
@@ -0,0 +1,210 @@
+import io.netty.handler.codec.dns.DatagramDnsQuery;
+import io.netty.handler.codec.dns.DatagramDnsResponse;
+import io.netty.handler.codec.dns.DefaultDnsQuestion;
+import io.netty.handler.codec.dns.DnsOpCode;
+import io.netty.handler.codec.dns.DnsQuery;
+import io.netty.handler.codec.dns.DnsQuestion;
+import io.netty.handler.codec.dns.DnsRawRecord;
+import io.netty.handler.codec.dns.DnsRecord;
+import io.netty.handler.codec.dns.DnsRecordDecoder;
+import io.netty.handler.codec.dns.DnsRecordEncoder;
+import io.netty.handler.codec.dns.DnsRecordType;
+import io.netty.handler.codec.dns.DnsResponse;
+import io.netty.handler.codec.dns.DnsResponseCode;
+import io.netty.handler.codec.dns.DnsSection;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.UnknownHostException;
+import java.util.ArrayList;
+import java.util.List;
+import okio.ByteString;
+
+public class DnsRecordCodec {
+  private static final InetSocketAddress DUMMY =
+      InetSocketAddress.createUnresolved("localhost", 53);
+
+  public static String encodeQuery(String host, boolean includeIPv6) throws Exception {
+    DatagramDnsQuery query = new DatagramDnsQuery(DUMMY, DUMMY, 0);
+    query.setRecursionDesired(true);
+    int index = 0;
+    query.addRecord(DnsSection.QUESTION, index++, new DefaultDnsQuestion(host, DnsRecordType.A));
+    if (includeIPv6) {
+      query.addRecord(DnsSection.QUESTION, index++, new DefaultDnsQuestion(host, DnsRecordType.AAAA));
+    }
+
+    //System.out.println("Query: " + query);
+
+    String encoded = encode(query);
+
+    //System.out.println("Query: " + encoded);
+
+    return encoded;
+  }
+
+  public static List<InetAddress> decodeAnswers(String hostname, ByteString byteString) throws Exception {
+    System.out.println("Response: " + byteString.hex());
+
+    DnsResponse response = decode(Unpooled.wrappedBuffer(byteString.asByteBuffer()));
+
+    //System.out.println("Response: " + response);
+
+    //System.out.println("Code: " + response.code());
+    if (response.code() == DnsResponseCode.NXDOMAIN) {
+      throw new UnknownHostException(hostname + ": NXDOMAIN");
+    } else if (response.code() == DnsResponseCode.SERVFAIL) {
+      throw new UnknownHostException(hostname + ": SERVFAIL");
+    }
+    // TODO check for SERVFAIL or NXDOMAIN
+
+    int recordCount = response.count(DnsSection.ANSWER);
+    List<InetAddress> result = new ArrayList<>(recordCount);
+    for (int i = 0; i < recordCount; i++) {
+      DnsRecord answer = response.recordAt(DnsSection.ANSWER, i);
+      if (answer.type() == DnsRecordType.A || answer.type() == DnsRecordType.AAAA) {
+        DnsRawRecord record = (DnsRawRecord) answer;
+        ByteBuf content = record.content();
+        byte[] bytes = new byte[content.readableBytes()];
+        content.readBytes(bytes);
+        result.add(InetAddress.getByAddress(bytes));
+      }
+    }
+    return result;
+  }
+
+  // Netty methods - TODO reimplement as minimal fixed without netty dependency
```

Comment:
Yes please

Reply:
Done, but it's the most awkward part of the code :(